### PR TITLE
Feature configparse 210401

### PR DESCRIPTION
**Pull request by mikaell**
Redoing the work from home…

### CODE REVIEW

**Comment by mikaell** [comment]
This Pull Request introduces usage of Pydantic for configuration parsing.

* Introduces automatic type checking for reading configs
* Improved readability
* Passes all our existing Pytests without modifying the tests.


This PR adds a functionality.

**How to test:**
Run pytests.

Test to load configurations.

**Expected outcome:**
Pytests should pass as usual.

Functionality should be working.


**Review:**
- [ ]  code approved by
- [ ]  tests executed by


Replaces pr # #2465

**Comment by northwestwitch** [comment]
Nice! We just merged another PR so the branch is not up tp date. We can review next week!

**Comment by northwestwitch** [comment]
We are in the process of preparing a new release for this week. I would like to avoid including this one in this release because I feel like this one a more extensive testing that we have time for at the moment. Also loading cases in a way that we don't do in prod (for instance using ped files etc). I would suggest let's review and merge it after the new release, do you agree @dnil and @mikaell ?

**Comment by mikaell** [comment]
> We are in the process of preparing a new release for this week. I would like to avoid including this one in this release because I feel like this one needs a more extensive testing that we have time for at the moment. Also loading cases in a way that we don't do in prod (for instance using ped files etc). I would suggest let's review and merge it after the new release, do you agree @dnil and @mikaell ?

Yes, let's not rush this PR. I'm about to start a review of the configparam documentation as well (as of some issue I don't remember which). These two should be added at the same time.

File: scout/parse/case.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 
 from scout.constants import PHENOTYPE_MAP, REV_PHENOTYPE_MAP, REV_SEX_MAP, SEX_MAP
 from scout.exceptions import ConfigError, PedigreeError
```
Comment (northwestwitch) [inline scout/parse/case.py:11]: ConfigError doesn't look used any more in this file:
```suggestion
from scout.exceptions import PedigreeError
```

File: scout/parse/case.py [hunk 2]
```diff
@@ -9,6 +9,7 @@
...
+from scout.parse.config_base import ScoutIndividual, ScoutLoadConfig, VcfFiles
```
Comment (northwestwitch) [inline scout/parse/case.py:12]: Same ScoutIndividual
```suggestion
from scout.parse.config_base import ScoutLoadConfig, VcfFiles
```

File: scout/parse/case.py [hunk 3]
```diff
@@ -25,7 +26,11 @@ def get_correct_date(date_info):
     Returns:
         correct_date(datetime.datetime)
     """
+    LOG.debug(date_info)
+    if isinstance(date_info, str):
```
Comment (northwestwitch) [inline scout/parse/case.py:30]: You forgot to remove something here

File: scout/parse/case.py [hunk 4]
```diff
@@ -41,131 +46,87 @@ def get_correct_date(date_info):
...
         vcf_str(str): Path to a VCF file
         vcf_sv(str): Path to a vcf file
-        vcf_cancer(str): Path to a vcf file
-        vcf_cancer_sv(str): Path to a vcf file
-        smn_tsv(str): Path to an SMN tsv file
-        peddy_ped(str): Path to a peddy ped
-        multiqc(str): Path to dir with multiqc information
-        cnv_report: Path to pdf file with CNV report
-        coverage_qc_report: Path to html file with coverage and qc report
-        gene_fusion_report: Path to the gene fusions report
```
Comment (northwestwitch) [inline scout/parse/case.py:84]: These last 2 items in the list were removed and not ordered alphabetically in this comment

File: scout/parse/case.py [hunk 5]
```diff
@@ -41,131 +46,87 @@ def get_correct_date(date_info):
...
+    if "ped" in kwargs and kwargs["ped"] is not None:
+        family_id, samples = parse_ped(kwargs["ped"])
+        config["family"] = family_id
+        config["samples"] = samples
+
+    # Give passed keyword arguments precedence over file configuration
+    # Except for 'owner', prededence config file over arguments
+    LOG.debug("KWARGS: {}".format(kwargs))
+    if "owner" in config:
+        print(kwargs)
```
Comment (northwestwitch) [inline scout/parse/case.py:87]: This print statement can be removed right?

File: scout/parse/case.py [hunk 6]
```diff
@@ -41,131 +46,87 @@ def get_correct_date(date_info):
...
+        if kwargs[key] is not None:
+            config[key] = kwargs[key]
+        else:
+            try:
+                config[key] = config[key]
+            except KeyError:
+                config[key] = None
+
+    # populate configuration according to Pydantic defined classes
+    LOG.debug("1st SCOUTLOADCONFIG")
```
Comment (northwestwitch) [inline scout/parse/case.py:99]: This can be removed

File: scout/parse/case.py [hunk 7]
```diff
@@ -609,3 +502,32 @@ def removeNoneValues(dictionary):
     """
 
     return {k: v for k, v in dictionary.items() if v is not None}
+
+
+def removeNoneRecursive(dictionary):
+    """Recusivly remove None from dictionary"""
```
Comment (northwestwitch) [inline scout/parse/case.py:508]:
```suggestion
    """Recusively remove None from dictionary"""
```

**Review by northwestwitch — CHANGES_REQUESTED** [review]
Nice start! 
I've started to review the code of this PR, but I've realized that the easiest way to start reviewing this PR would be to first compare the results you get when loading a case using master branch and this branch.

Using `scout setup demo` command:

Loading case using master:

image

This branch:
image

------
Master:
image

This branch is missing sex check and parenthood:
image


------
Master:
image

This branch:

image

------

Master:

image

This branch:
image

------

Case sidebar of this branch (on the right) is not collapsible and it's missing "gene panel":

image


I made a few comments in the code and I think it would be nice starting to fix these things, the we could give it another go!

**Comment by mikaell** [comment]
Thanks for the review! I've been preoccupied lately and haven't had the time work on this. I'll be on CG's 100% from June.

**Comment by northwestwitch** [comment]
> Thanks for the review! I've been preoccupied lately and haven't had the time work on this. I'll be on CG's 100% from June.

Sorry to hear that, no worries!

**Comment by northwestwitch** [comment]
Sorry this PR is not passing a test any more after we've introduced a couple of new keys in the case config file. Check this PR to see the introduced changes: #2659

**Comment by northwestwitch** [comment]
@mikaell let me know when you are ready for another review of this one!

File: scout/parse/case.py (hunk 1 above)
Comment (mikaell) [reply scout/parse/case.py:11]: Fixed!

**Comment by northwestwitch** [comment]
@mikaell I'll review this one first thing tomorrow. You could try to fix the codefactor issues in the meanwhile!

File: scout/parse/config_base.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+                if father not in all_ids:
+                    raise PedigreeError("father %s does not exist in family" % father)
+            mother = parsed_ind.get("mother")
+            if mother and mother != "0":
+                if mother not in all_ids:
+                    raise PedigreeError("mother %s does not exist in family" % mother)
+        return individuals
+
+    @validator("track")
+    def field_not_none(cls, v):
```
Comment (northwestwitch) [inline scout/parse/config_base.py:215]: I don't understand why has validator("track") and then it says "Owner and family can not be None"

File: scout/parse/config_base.py [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+        return value
+
+    @validator("family", pre=True, always=True)
+    def mandatory_check_family(cls, value):
+        if value is None:
+            raise ConfigError("A case has to have a 'family'")
+        return value
+
+    @validator("madeline_info")
+    def check_if_madeline_exists(cls, path):
```
Comment (northwestwitch) [inline scout/parse/config_base.py:188]: Doesn't look like this one is working. The pedigree is missing in the demo case.. 🤔

File: scout/parse/config_base.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
+"""Class to hold information about scout load config"""
```
Comment (northwestwitch) [inline scout/parse/config_base.py:1]: Perhaps renaming this file to something more self-explanatory, for instance "case_config_base" ?

File: scout/parse/case.py [hunk 11]
```diff
@@ -32,140 +29,90 @@ def get_correct_date(date_info):
...
+        if kwargs[key] is not None:
+            config[key] = kwargs[key]
+        else:
+            try:
+                config[key] = config[key]
+            except KeyError:
+                config[key] = None
+
+    # populate configuration according to Pydantic defined classes
+    LOG.debug("1st SCOUTLOADCONFIG: {}".format(config))
```
Comment (northwestwitch) [inline scout/parse/case.py:92]: What is 1st SCOUTLOADCONFIG?

File: scout/parse/config_base.py [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+    def check_if_madeline_exists(cls, path):
+        """Add the pedigree figure, this is a xml file which is dumped in the db"""
+        mad_path = Path(path)
+        if not mad_path.exists():
+            raise ValueError("madeline path not found: {}".format(mad_path))
+        with mad_path.open("r") as in_handle:
+            return in_handle.read()
+
+    @validator("individuals")
+    def family_relations_consistent(cls, individuals):
```
Comment (northwestwitch) [inline scout/parse/config_base.py:197]: Nice!

File: scout/parse/config_base.py [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+
+    # Each case has to have a owner. If not provided in config file it needs to be given as a
+    # argument
+    @validator("owner", pre=True, always=True)
+    def mandatory_check_owner(cls, value):
+        if value is None:
+            raise ConfigError("A case has to have a owner")
+        return value
+
+    @validator("family", pre=True, always=True)
```
Comment (northwestwitch) [inline scout/parse/config_base.py:181]: Consider adding a stringdoc for each one of these function to describe shortly what they do, for future reference and  maintenance

File: scout/parse/config_base.py [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+                or vcf_dict["vcf_cancer_research"]
+                or vcf_dict["vcf_cancer_sv"]
+                or vcf_dict["vcf_cancer_sv_research"]
+            ):
+                values.update({"track": "cancer"})
+        except Exception as error:
+            LOG.error("exception in vcf_s! Not set? {}".format(error))
+
+        # Update collaborators to [owner] if not set
+        if values.get("collaborators") is None:
```
Comment (northwestwitch) [inline scout/parse/config_base.py:247]: Why is collaborator settings inside the VCF track validator? These fields are not related

File: scout/parse/config_base.py [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+        return v
+
+    @validator("synopsis")
+    def synopsis_pre_hook(cls, my_synopsis):
+        if isinstance(my_synopsis, list):
+            return ". ".join(my_synopsis)
+        return my_synopsis
+
+    @root_validator
+    def update_track(cls, values):
```
Comment (northwestwitch) [inline scout/parse/config_base.py:227]: Rename to update_cancer_vcf_tracks?

File: tests/commands/load/test_load_case_cmd.py [hunk 16]
```diff
@@ -116,25 +98,19 @@ def test_load_case_SyntaxError(mock_app, institute_obj, case_obj, monkeypatch):
     def mock_parse_case(*args):
```
Comment (northwestwitch) [inline tests/commands/load/test_load_case_cmd.py:98]: I think it would be much more useful for testing purposes is the error is naturally triggered by a faulty config file, instead of a mocker that triggers the error like this

**Review by northwestwitch — CHANGES_REQUESTED** [review]
Reviewed the code and noticed how many line it is possible to remove, nice! I have some suggestion on how to improve, see my comments.

Tested it and there are still problems unfortunately, check also my [previous comment from april](https://github.com/Clinical-Genomics/scout/pull/2496#pullrequestreview-645852996):
- buttons to VCF files work now 🆗 
- pedigree is still missing ⛔ 
- Sex check is still missing ⛔ 
- Parenthood confirmation is still missing ⛔ 
- Gene panels are loaded now 🆗 
- Sidebar is still not collapsible and missing information: ⛔ 

image

Looks like that still not all necessary keys/values are collected and the right keys created in the database for the case object. That's why I've proposed to introduce tests that check that the created case contains all the expected keys, before or after the case is saved to database..

File: scout/parse/case.py [hunk 17]
```diff
@@ -612,3 +317,35 @@ def removeNoneValues(dictionary):
     """
 
     return {k: v for k, v in dictionary.items() if v is not None}
+
+
+def remove_none_recursive(dictionary):
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/parse/case.py:338]: Naming a assisting function `..._aux` is common for non-tail recursive assisting function.

File: scout/parse/case.py [hunk 18]
```diff
@@ -612,3 +317,35 @@ def removeNoneValues(dictionary):
...
+    """Recusivly remove None from dictionary"""
+    return remove_none_recursive_aux(dictionary, {})
+
+
+def remove_none_recursive_aux(dictionary, new_dict):
+    """Auxilary Function to remove_None_recursive"""
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/parse/case.py:344]: Done!

File: scout/parse/case.py [hunk 19]
```diff
@@ -612,3 +317,35 @@ def removeNoneValues(dictionary):
...
+    for key, value in dictionary.items():
+        if value is not None:
+            new_dict.update({key: value})
+        if (
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/parse/case.py]: I've elaborated docstring, I think that should be sufficient.

File: scout/parse/config_base.py (hunk 13 above)
Comment (mikaell) [reply scout/parse/config_base.py]: Done!

File: tests/commands/load/test_load_case_cmd.py (hunk 16 above)
Comment (mikaell) [reply tests/commands/load/test_load_case_cmd.py:98]: This error is mostly hypothetical. Any syntax error in the configuration will be caught by the yaml parser and secondly by the pydantic parser. I haven't come up with a way to manipulate the config file to get through.

**Comment by mikaell** [comment]
Collapse now works.

File: scout/parse/config_base.py [hunk 20]
```diff
@@ -0,0 +1,262 @@
+"""Class to hold information about scout load config"""
+
+import copy
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/parse/config_base.py]: Removed!

File: scout/parse/case.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/parse/case.py:348]: I found rewriting the function hard, and not gaining readability (in my opinion). Is it ok to keep as it is?

**Comment by mikaell** [comment]
I think this is ready for another review.

**Comment by northwestwitch** [comment]
@mikaell unfortunately there are a couple of conflicts now. Could you fix them when you have time, so we can review it once more!

**Comment by mikaell** [comment]
> @mikaell unfortunately there are a couple of conflicts now. Could you fix them when you have time, so we can review it once more!

Ok. I'll have a look.

**Comment by mikaell** [comment]
This is ready for one more review!

**Review by northwestwitch — APPROVED** [review]
It's working well now. I've tested it only locally with the demo case + the demo cancer case and they both look fine.

I've got a few small suggestions but it's fine!

Since this PR is about fixing case parsing and creation this might be a good place to remove those 2 extra keys present and not used in the case document:
- chromograph_image_files
- chromograph_prefixes

There is an open issue about it.

I'm approving but **PLEASE DON*t MERGE BEFORE WE TEST IT WITH REAL CASES ON STAGE**

PS: I' cant believe we are almost there with this one 😆

File: scout/parse/case.py [hunk 21]
```diff
@@ -32,178 +32,112 @@ def get_correct_date(date_info):
...
-        config_data["default_gene_panels"] = [
-            panel.strip() for panel in config_data["default_gene_panels"]
-        ]
+    LOG.debug("KWARGS: {}".format(kwargs))
+    config = kwargs.pop("config", {})
 
-    # handle scout/commands/load/case.py
-    config_data["custom_images"] = parse_custom_images(config_data)
+    # populate configuration according to Pydantic defined classes
+    LOG.debug("1st SCOUTLOADCONFIG: {}".format(config))
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/parse/case.py:75]: Old debug print (I used to keep track of function call order). Will be removed.

File: scout/parse/models.py [hunk 22]
```diff
@@ -0,0 +1,320 @@
...
+
+    @validator("custom_images")
+    def remove_empty_images(cls, custom_images):
+        LOG.debug("CUSTOM: {}".format(custom_images))
+
+        for section in custom_images:
+            images = custom_images[section]
+            LOG.debug("IMAGES: {}".format(images))
+            for i in images:
+                LOG.debug("III: {}".format(i))
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/parse/models.py]: Removed!

File: scout/parse/case.py [hunk 23]
```diff
@@ -522,89 +281,23 @@ def parse_case(config):
...
-    if "owner" not in config:
-        raise ConfigError("A case has to have a owner")
-
-    if "family" not in config:
-        raise ConfigError("A case has to have a 'family'")
-
-    individuals = parse_individuals(config["samples"])
-    synopsis = None
+    # create a config object based on pydantic rules
+    LOG.debug("*** NO LONGER CALLED ***")
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/parse/case.py]: Removed!

File: scout/parse/case.py [hunk 24]
```diff
@@ -315,204 +248,30 @@ def add_peddy_information(config_data):
...
-        mother = parsed_ind.get("mother")
-        if mother and mother != "0":
-            if mother not in ind_ids:
-                raise PedigreeError("mother %s does not exist in family" % mother)
-
-    return individuals
+                    analysis_inds[parent_id]["confirmed_parent"] = True
 
 
+# XXX: No longer called! (previously called from mongo/case)
```
[the remark this answers is not in the document]
Comment (mikaell) [reply scout/parse/case.py]: Removed.
